pkg/keycloak/realm/phaseHandler.go

```diff
@@ -0,0 +1,449 @@
+	for i := range realm.Spec.IdentityProviders {
+		identityProvider := &realm.Spec.IdentityProviders[i]
+		identityProviderPairsList[identityProvider.Alias] = &v1alpha1.KeycloakIdentityProviderPair{
+			SpecIdentityProvider: identityProvider,
+			KcIdentityProvider:   kcIdentityProviders[identityProvider.Alias],
+		}
+		delete(kcIdentityProviders, identityProvider.Alias)
+	}
+
+	for i := range kcIdentityProviders {
+		identityProvider := kcIdentityProviders[i]
+		identityProviderPairsList[identityProvider.Alias] = &v1alpha1.KeycloakIdentityProviderPair{
+			KcIdentityProvider:   identityProvider,
+			SpecIdentityProvider: nil,
+		}
+	}
+
+	for i := range identityProviderPairsList {
+		err := ph.reconcileIdentityProvider(identityProviderPairsList[i].KcIdentityProvider, identityProviderPairsList[i].SpecIdentityProvider, realm.Spec.Realm, kcClient)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ph *phaseHandler) reconcileIdentityProvider(kcIdentityProvider, specIdentityProvider *v1alpha1.KeycloakIdentityProvider, realmName string, authenticatedClient keycloak.KeycloakInterface) error {
+	if specIdentityProvider == nil {
+		err := authenticatedClient.DeleteIdentityProvider(kcIdentityProvider.Alias, realmName)
+		if err != nil {
+			return err
+		}
+	} else if kcIdentityProvider == nil {
+		err := authenticatedClient.CreateIdentityProvider(specIdentityProvider, realmName)
+		if err != nil {
+			return err
+		}
+	} else {
+		//The API doesn't return the secret, so in order to stop in never being equal we just set it to the spec version
+		kcIdentityProvider.Config["clientSecret"] = specIdentityProvider.Config["clientSecret"]
+		//Ensure the internalID is set on the spec object, this is required for update requests to succeed
+		specIdentityProvider.InternalID = kcIdentityProvider.InternalID
+		if !resourcesEqual(kcIdentityProvider, specIdentityProvider) {
+			err := authenticatedClient.UpdateIdentityProvider(specIdentityProvider, realmName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (ph *phaseHandler) Deprovision(realm *v1alpha1.KeycloakRealm) (*v1alpha1.KeycloakRealm, error) {
+	kcClient, err := ph.getClient(realm)
+	if err != nil {
+		return realm, err
+	}
+
+	//delete client secrets
+	for _, client := range realm.Spec.Clients {
+		clientSecret := &corev1.Secret{
+			TypeMeta: v1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Secret",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Namespace: realm.GetNamespace(),
+				Name:      *client.OutputSecret,
+			},
+		}
+		ph.sdk.Delete(clientSecret)
+	}
+
+	//delete user secrets
+	for _, user := range realm.Spec.Users {
+		userSecret := &corev1.Secret{
+			TypeMeta: v1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Secret",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Namespace: realm.GetNamespace(),
+				Name:      *user.OutputSecret,
+			},
+		}
+		ph.sdk.Delete(userSecret)
+	}
+
+	err = kcClient.DeleteRealm(realm.Spec.ID)
```

Comment:
this is cleaning up the deletion of the keycloak realm CR. So the keycloak instance could still be running for other realms, or for the recreation of this realm? I think cleaning out the realm from the instance when the realm CR is deleted makes sense.

Reply:
ok that makes sense. Do all of the Realms naturally get cleaned up when the parent CR is deleted and is it that done via an owner ref?

Reply:
I've added some code to check that the keycloak API can be reached before trying to operate on a keycloak realm CR that has a keycloak instance assigned to it. If it fails it will leave a message on the CR.

I think just nuking the CR when a seperate CR is deleted could upset some devs if they've spent a while tweaking it?

Reply:
> I think just nuking the CR when a seperate CR is deleted could upset some devs if they've spent a while tweaking it?
I assume this CR would be checked into version control so I don't see that being an issue.
I am not sure we would want it auto assigned to another instance either. 
Might be better at this point to add a status if the backing keycloak CR is removed?

Reply:
sure, so on delete of a keycloak CR, find all they keycloak realm CRs that are using it, and set the status on them?